### DISCUSSION TITLE
Proposal to fix Staking contract for new staker

### DISCUSSION THREAD

**Discussion by kfujita0520** [pavlovcik/uad-ui#369] [Protocol Specifications]
Currently Staking contract in production has some issue. When user tried to withdraw their LP token which they have staked after specific period, they could only receive much less amount than they have staked.
While most of stakers might be internal memeber, I think this is serious problem, because when any new comer tried to stake their LP token for testing purpose in short term, they will have very bad experience as they cannot get back their LP tokens after staking period, which may cause negative impression to project itself.

In my observation, the technial problem right now is, the system calculated returned amount in wired way (return proportional amount to actual holding amount in BondingV2 contract). But because you would not hold all deposited LP tokens in Bonding V2 contract, reported issue will happen.
I believe we can fix the code to return the original amount as it should be and if requested withdrawal amount is more than you hold, system can throw error.  In this way, you could avoid the issue. If you could stop withdrawal of big amount staker, which is usually internal member having longer stake period, this issue won't happen actually. 

To minimize the impact on existing system, I propose we can just roll out new BondingFormula contract and switch it by updating the address of management contract. 
If this proposal is fine for you, I can work on that. Your kind attention and advise is highly appreciated.

**Comment by 0x4007** [comment]
I think @zgorizzo69 might know the best on whether this proposed solution is viable. 

You bring up a good point about newcomers probably having a bad feeling about the project. I think a more immediate fix could be to have a warning message on the UI next to the withdraw buttons that say to contact us for a reimbursement of the difference if you are withdrawing.

**Comment by zgorizzo69** [comment]
first thing first ..
can you write a test demonstrating what is wrong with the calculation ? 
then we can use this test to assess that the fix solves the problem

**Reply by kfujita0520** [reply]
@zgorizzo69 
Thank you for your message. people could not receive all LP tokens they have staked. So they will lose LP tokens after staking period. That is the current problem. 
If you need the detail transaction example, i can share it my private transaction(tx id) in DM of telegram.

**Reply by 0x4007** [reply]
To clarify, the math seems to be correct but the issue is that we pulled out liquidity from the market. 

I believe that @kfujita0520 is proposing to modify the math to keep the staked amounts and their withdrawals amount constant/identical instead of a percentage of the pool. 

Pros: solves our immediate issue with newcomers staking for a small bit of time and losing money.
Cons: if we leave this in, we could have bank run type scenarios occur. 

Currently for the four year stakers, they still have to wait approximately 148 weeks as of now to withdraw which means that if we can reimburse the liquidity here from the liquidity we plan to raise from our Bonds then everything will be fine.
